libafl/src/events/hooks/mod.rs

```diff
@@ -0,0 +1,118 @@
+//! Hooks for event managers, especifically these are used to hook before and and `handle_in_client`.
+//! This will allow user to define pre/post-processing code when the event manager receives any message from
+//! other clients
+use libafl_bolts::ClientId;
+
+use crate::{events::Event, state::State, Error};
+
+/// The hooks that are run before and after the event manager calls `handle_in_client`
+pub trait EventManagerHook<S>
+where
+    S: State,
+{
+    /// The hook that runs before `handle_in_client`
+    /// Return false if you want to cancel the subsequent event handling
+    fn pre_exec<E, Z>(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        client_id: ClientId,
+        event: &Event<S::Input>,
+    ) -> Result<bool, Error>;
+    /// The hook that runs after `handle_in_client`
+    /// Return false if you want to cancel the subsequent event handling
+    fn post_exec<E, Z>(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        client_id: ClientId,
+    ) -> Result<bool, Error>;
+}
+
+/// The tuples contains hooks to be executed for `handle_in_client`
+pub trait EventManagerHooksTuple<S>
+where
+    S: State,
+{
+    /// The hook that runs before `handle_in_client`
+    fn pre_exec_all<E, Z>(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        client_id: ClientId,
+        event: &Event<S::Input>,
+    ) -> Result<bool, Error>;
+    /// The hook that runs after `handle_in_client`
+    fn post_exec_all<E, Z>(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        client_id: ClientId,
+    ) -> Result<bool, Error>;
+}
+
+impl<S> EventManagerHooksTuple<S> for ()
+where
+    S: State,
+{
+    /// The hook that runs before `handle_in_client`
+    fn pre_exec_all<E, Z>(
+        &mut self,
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        _state: &mut S,
+        _client_id: ClientId,
+        _event: &Event<S::Input>,
+    ) -> Result<bool, Error> {
+        Ok(true)
+    }
+    /// The hook that runs after `handle_in_client`
+    fn post_exec_all<E, Z>(
+        &mut self,
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        _state: &mut S,
+        _client_id: ClientId,
+    ) -> Result<bool, Error> {
+        Ok(true)
+    }
+}
+
+impl<Head, Tail, S> EventManagerHooksTuple<S> for (Head, Tail)
```

Comment:
The old question: why don't we just do a `vec` with the hooks in it?

Reply:
Cannot constrain per-type/cannot look for specific trait impls/etc.

Reply:
just give it all we got

Reply:
Adding and removing hooks on the fly will be useful

Reply:
I just don't understand why one wouldn't implement a hook with dynamic behaviour instead. You aren't going to be generating hooks.

Reply:
can't do this because pre_exec/post_exec have generics <E, Z>

Reply:
FnMut<E, Z> can have generics too

Reply:
i mean i can't make them into vec because the compiler says no because it has generics

Reply:
Shoudn't `Vec<Box<dyn FnMut<E, Z>(&blubb) -> bla>>` work?

Reply:
You can't make a generic dyn. You can have a concretised generic type in a dyn, but you cannot have dyns with generic arguments.

Reply:
They are concrete here, right? If they are in an `struct blubb<E, Z>`

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bca6fdf9ae5fa1aa85e0193a6cd47c8d

This is what I meant

Reply:
That's not the situation... The function being called must be generic. There's no concrete type A or B beforehand.

Reply:
Ah I get what you mean... The obvious solution would be not to give the hooks access to executor or fuzzer

Reply:
In any case, access to state and event manager should be fine tbh